File: #1. 모의고사 2회/장효빈/problem3_카페_확장.py
# 풀이 시간: 35분 
# 시간 복잡도 O(N)
# 코드 설명: 먼저 온 손님의 음료를 먼저 제조하므로 queue 자료구조를 선택하여 품
# 취약한 부분: 큐를 떠올리는 것?

from collections import deque

def solution(menu, order, k):
    q = deque()
    answer = 0
    for o in order:
        time = k
        q.append(menu[o])
        answer = max(answer, len(q))
        while q and time > 0:
            if q[0] <= time:
                time -= q[0]
                q.popleft()
            else:
                q[0] -= time
                time = 0
        
    return answer

File: #1. 모의고사 2회/김민정/problem3_카페확장.java
// 풀이 시간: 50분
// 시간 복잡도: O(number*logN + N)
// 코드 설명: k초마다 카페에 사람이 얼마나 있는지 확인한다
// 취약한 부분: 'k초마다' 확인해도 된다는 걸 떠올리는데 시간이 오래 걸림

import java.util.*;

class Solution {
    public int solution(int[] menu, int[] order, int k) {
        int answer = Integer.MIN_VALUE;

        Map<Integer, Integer> cnt = new HashMap<>();
        int prev = 0;
        for (int i = 0; i < order.length; i++) {
            int orderNum = order[i];

            int start = i * k;
            int end = Math.max(prev, start) + menu[orderNum];

            for (int j = start; j < end; j+=k) {
                cnt.put(j, cnt.getOrDefault(j, 0) + 1);
                answer = Math.max(answer, cnt.get(j));
            }

            prev = end;
        }

        return answer == Integer.MIN_VALUE ? 1 : answer;
    }
}